### PR TITLE
feat(acms parser): Adds a parser class to handle attachment pages.

### PR DESCRIPTION
**Pull request by ERosendo**
Adds support for parsing attachment pages coming from ACMS.

### CODE REVIEW

**Comment by mlissner** [comment]
@johnhawkinson any interest/time/willingness to review this one? I'm still recovering from being sick and I think you know this part as well as anybody at this point?

**Comment by johnhawkinson** [comment]
Yeah, I'll try to look at this early tomorrow morning.

**Comment by mlissner** [comment]
@johnhawkinson do you still want to take a look at this? I think Eduardo has filed all his PRs, so we need to review things and we'll have ACMS once everything is merged. If you've gotten busy, somebody over here can look too. I thought I'd ask since you've got a better sense of the data.

**Comment by johnhawkinson** [comment]
My bad I got distracted with Thanksgiving and failed to come back. I'll try to take a good look at this shortly.

**Comment by mlissner** [comment]
Pinging this for you, @johnhawkinson, because I know you were pinging us for the ACMS stuff. :)

**Comment by grossir** [comment]
@ERosendo This looks great on the Juriscraper side. I would suggest giving the example files a JSON extension, and formatting them properly so they are readable 

To understand this PR properly, I also took a look at the related courtlistener [PR](https://github.com/freelawproject/courtlistener/pull/3393/files). I left some comments there.

**Comment by ERosendo** [comment]
@grossir thanks for the review. I've applied your suggestions about the example files. 👍

**Comment by ERosendo** [comment]
> Is this really a good idea

I believe changing the file extension to `.json` is a good idea. My IDE was able to recognize the file format immediately and suggested proper formatting. This is an improvement over the previous `.acms_json` extension, which treated the file as plain text. Consistent formatting will definitely benefit everyone collaborating on the project.

> does it actually work?

Yes, It does. 

Your comment regarding the `PacerParseTestCase` class is correct. While it expects output files with the `.json` extension, the input files can have various extensions like `.acms.json` or `.compare.json`. This is because the `parse_files` method discards any extensions during processing and then attempts to compare the input file with another file that shares the same base name but has the `.json` extension.

https://github.com/freelawproject/juriscraper/blob/9222abd0a2755384de25d3fb17ecd6c05263b760/tests/local/PacerParseTestCase.py#L27-L28

File: tests/examples/pacer/appellate_attachment_pages/acms/ca2_23-7246_003.compare.json
Comment (mlissner) [inline tests/examples/pacer/appellate_attachment_pages/acms/ca2_23-7246_003.compare.json:1]: There are a LOT of fields in here. We should figure out which make sense to include in the returned object, and which are new. It feels foolish to throw away the data we get just because in district court we don't get the same data, right?

**Review by mlissner — CHANGES_REQUESTED** [review]
Only two comments from me, but one will take some work, sorry. I'm afraid now is probably the right time to at least analyze what we're throwing away...

Thank you all!

File: tests/examples/pacer/appellate_attachment_pages/acms/ca2_23-7222_003.json
Comment (ERosendo) [reply tests/examples/pacer/appellate_attachment_pages/acms/ca2_23-7222_003.json:6]: This is the source for that output:

image

It's coming from entry 3 of the case [23-7222](https://ca2-showdoc.azurewebsites.us/23-7222).

File: tests/examples/pacer/appellate_attachment_pages/acms/ca2_23-7246_003.compare.json
Comment (ERosendo) [reply tests/examples/pacer/appellate_attachment_pages/acms/ca2_23-7246_003.compare.json:1]: You're right. While the elements within the `caseDetails` key are predominantly related to the docket and handled by the `ACMSDocketReport` class when the extension uploads the docket report, I included this key to extract the `caseId`. Extracting and including only the `caseId` in the request is a viable option, and the extension itself can handle that cleaning process.

The JSON from ACMS also provides valuable data for the case entry itself (`docketEntry` key) and each document associated with the attachment page (`docketEntryDocuments` key). here's a summary of the fields we got from the sessionStorage:

- docketEntry ( the `ACMSDocketReport` class also parses this data when the docket report is uploaded, [here's the code](https://github.com/freelawproject/juriscraper/blob/9222abd0a2755384de25d3fb17ecd6c05263b760/juriscraper/pacer/acms_docket.py#L289-L377))
  - `createdOn`: Date and time the entry was created.
  - `docketEntryId`: Unique identifier for the entry.
  - `docketEntryText`: Description of the entry.
  - `documentCount`: Total number of documents attached to the entry.
  - `endDate`: Date the entry was closed.
  - `endDateFormatted`: Human-readable format of the endDate.
  - `entryNumber`: Sequential number assigned to the entry within the docket.
  - `fileSize`: Total size of all attached documents in bytes.
  - `pageCount`: Total number of pages across all attached documents.
  - `restrictedDocsAvailable` (Boolean): 
   - `restrictedPartyFilingDocketEntry` (Boolean):


- docketEntryDocuments: each element inside this array represents a document attached to the docket entry.

   - `billablePages`: Number of pages considered billable (we're returning this as the `page_count`)
   - `caseFilingDocumentUrl`: This seems to be an Internal URL for the document within sharepoint. I tried clicking this link but I got the following error:
   
   
     image

   - `cost`:  Cost associated with the document. I think we can compute this value using the `page_count` but we could include it in the output. 
  
   - `createdOn`: Date and time the document was uploaded.
   
   - `docketDocumentDetailsId`: Unique identifier for the document. **We'll store this value in the new acms field introduced in** https://github.com/freelawproject/courtlistener/pull/3193.
   
   - `documentNumber`: Sequential number assigned to the document within the entry. **We're returning this as the `attachment_number`.**
  
   - `documentPermission`: It seems this key provides the permission level for accessing the document (?). 
   
   - `documentUrl`: This is another sharepoint URL.
  
   - `fileSize`: Size of the document in bytes. **This value is not included in the output but we could return it as `file_size`**.
   
   - `name`: Name of the entry. **We're returning this as `description`**
  
   - `searchTransaction` & `searchValue`: Likely for internal searching (?)

Let me know what you think.

Comment (mlissner) [reply tests/examples/pacer/appellate_attachment_pages/acms/ca2_23-7246_003.compare.json:1]: Cool, thanks. 

I think we should add these fields to our object:

| ACMS field | Where in our object | CourtListener DB | Notes |
|--------|--------|--------|-----|
| docketEntry.createdOn | root.date_filed | DocketEntry.date_filed | |
| docketEntry.docketEntryText | root.description | DocketEntry.description | | 
| docketEntry.endDate  | date_end | None | I don't know what this means, but Juriscraper can return it for now. | 
| docketEntryDocuments.cost | attachments.cost | None | This might be useful to distinguish opinions (which are free) from other docs |
|docketEntryDocuments.createdOn | attachments.date_filed | None | We don't keep dates for individual documents in our model; maybe we should, but until then we can at least have it in Juriscraper |
| docketEntryDocuments.documentPermission | attachments.permission | None | We'll see if we can learn what kinds of values we get here, and then start making use of them. |
| docketEntryDocuments.fileSize | attachments.file_size | RECAPDocument.file_size | Good info, let's grab it! |
| docketEntryDocuments.name | attachments.description | RECAPDocument.description | This needs to be cleaned up. I think I put that in an earlier comment. |

That should get us the best of the data here.

File: tests/examples/pacer/appellate_attachment_pages/acms/ca2_23-7222_003.json
Comment (mlissner) [reply tests/examples/pacer/appellate_attachment_pages/acms/ca2_23-7222_003.json:6]: Yeah, that's not right. this should be parsed down to better values. All we want in this field is "Case OPENED."

**Comment by ERosendo** [comment]
@mlissner I've updated the ACMSAttachmentPage class's output object to provide more comprehensive data. Additionally, I've introduced a new helper method to ensure the descriptions of entries on the attachment page are sanitized.

File: tests/examples/pacer/appellate_attachment_pages/acms/ca2_23-7222_003.json
Comment (mlissner) [reply tests/examples/pacer/appellate_attachment_pages/acms/ca2_23-7222_003.json]: Sorry, looks like this one still isn't resolved?

Comment (ERosendo) [reply tests/examples/pacer/appellate_attachment_pages/acms/ca2_23-7222_003.json]: I addressed this comment in commit  https://github.com/freelawproject/juriscraper/pull/780/commits/38609e0aee3e1c97dffe3ddbd265887569b2d19e, Do you think it still needs work?

**Comment by mlissner** [comment]
Sorry, I must have missed your changes. Setting to auto-merge! Thank you.

**Comment by ERosendo** [comment]
Hey @mlissner, we need your approval to merge this PR 😄

**Review by mlissner — APPROVED** [review]

**Comment by mlissner** [comment]
Woo! Another one from Nov. 2023 is merged!

**Pull request merged**
